# Porównanie z dokładnym rozwiązaniem
W celu znalezienia dokładnego rozwiązania rozważyliśmy prosty przypadek o następujących danych:

* dni zbiorów: 2
* pojemność magazynu: 1
* dzienny limit zbiorów: 2
* 2 typy owoców o następujących danych
    1. Gruszki:
        - kilogramy w sadzie: 3
        - koszt zasadzenia: 1.5
        - ceny podstawowe na targu: [2, 1.5]
        - ceny w skupie: [1, 1]
        - popyt: [1, 3]
    2. Jabłka:
        - kilogramy w sadzie: 3
        - koszt zasadzenia: 2
        - ceny podstawowe na targu: [2, 3]
        - ceny w skupie: [1, 1.5]
        - popyt: [2, 2]
* mnożnik ceny:
    ```
    def multiplier(k):
        if k < 30:
            return 1.5
        if k < 50:
            return 1.2
        if k < 80:
            return 1.1
        if k <= 100:
            return 1
     ```
* koszty pracowników:
    ```
    def employee_cost(kilograms):
        if 0 <= kilograms <= 1:
            cost = 1.5
        elif 1 < kilograms <= 2:
            cost = 2
        else:
            cost = 3
        return cost
    ```
* koszty magazynowania:
    ```
    def warehouse_cost(kilograms):
        if 0 <= kilograms <= 1:
            cost = 1
        elif 1 < kilograms <= 2:
            cost = 2
        else:
            cost = 3
        return cost
    ```
 
Następnie ręcznie wprowadzono wszystkie sposoby na jakie w ciągu dnia można ułożyć zbiory, sprzedaż i magazyn z założeniem, że nie posiadamy dodatkowych owoców z magazynu z dnia poprzedniego. Następnie w pętli po tych dziennych rozwiązaniach dodano wszystkie opcje z założeniem, że posiadamy jakieś owoce w magazynie z poprzedniego dnia.

Kolejnym krokiem było sprawdzenie wszystkich rozwiązań rozwiązań metodą brute force i wybranie tych, które spełniały ograniczenia.

In [1]:
import os
import sys
import json
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
from bruteforce_solution import main, warehouse_cost, multiplier, employee_cost
from project_app.solution_classes import FruitTypeInfo
from project_app.sad import Orchard

In [2]:
best_solutions, best_profit, valid_solutions = main()

Day 1
harvested: [1, 1]
sold_market: [1, 1]
sold_wholesale: [0, 0]
warehouse: [0, 0]

Day 2
harvested: [0, 2]
sold_market: [0, 2]
sold_wholesale: [0, 0]
warehouse: [0, 0]





Zysk: 0.7000000000000002
Poprawne rozwiązania: 1237
Ilość dziennych rozwiązań: 125


Powyżej widać, że nawet dla tak prostych założeń istniało aż 1237 rozwiązań spełniających warunki.

In [3]:
gruszki = FruitTypeInfo(name="gruszki", quantity=3, planting_cost=1.5, base_price=[2,1.5],
                        wholesale_price=[1,1], demand=[1,3], min_market_sold=1,
                        multiplier=multiplier)

jablka = FruitTypeInfo(name="jabłka", quantity=3, planting_cost=2, base_price=[2,3],
                        wholesale_price=[1,1.5], demand=[2,2], min_market_sold=12,
                        multiplier=multiplier)

orchard = Orchard(fruit_types=[gruszki, jablka], employee_cost=employee_cost,
                  warehouse_cost=warehouse_cost, max_daily_harvest=2, warehouse_capacity=1, num_days=2)

In [9]:
# Algorytm genetyczny, który w rozwiązaniach początkowych nie zawiera optymalnej strategii zbiorów.
strategies = [
                [[1,[1,1]],[1,[1,1]]],
                [[1,[2,0]],[1,[0,2]]],
                [[1,[1,0]],[1,[0,1]]],
                [[1,[0,1]],[1,[1,0]]],
                [[1,[0,0]],[1,[2,0]]],
                [[1,[1,1]],[1,[2,0]]],
                [[1,[2,0]],[1,[1,1]]],
                [[1,[0,2]],[1,[1,1]]]
]
sol, profit, iterations = orchard.genetic_algorithm(max_iter_no_progress=300, max_iter=3000,
                                            replacement_rate=0.5,
                                            mutation_proba=0.4, random_demand_rate=False, verbose=True,
                                                   bruteforce_comapre=strategies)

print(f"Wynik algorytmu genetycznego\n{sol}Zysk: {profit}")

best profit: 0.2999999999999998 | iteration number: 1
best profit: 0.2999999999999998 | iteration number: 2
best profit: 0.2999999999999998 | iteration number: 3
best profit: 0.2999999999999998 | iteration number: 4
best profit: 0.2999999999999998 | iteration number: 5
best profit: 0.2999999999999998 | iteration number: 6
best profit: 0.2999999999999998 | iteration number: 7
best profit: 0.2999999999999998 | iteration number: 8
best profit: 0.2999999999999998 | iteration number: 9
best profit: 0.2999999999999998 | iteration number: 10
best profit: 0.2999999999999998 | iteration number: 11
best profit: 0.2999999999999998 | iteration number: 12
best profit: 0.2999999999999998 | iteration number: 13
best profit: 0.2999999999999998 | iteration number: 14
best profit: 0.2999999999999998 | iteration number: 15
best profit: 0.2999999999999998 | iteration number: 16
best profit: 0.2999999999999998 | iteration number: 17
best profit: 0.2999999999999998 | iteration number: 18
best profit: 0.2999

best profit: 0.2999999999999998 | iteration number: 219
best profit: 0.2999999999999998 | iteration number: 220
best profit: 0.2999999999999998 | iteration number: 221
best profit: 0.2999999999999998 | iteration number: 222
best profit: 0.2999999999999998 | iteration number: 223
best profit: 0.2999999999999998 | iteration number: 224
best profit: 0.2999999999999998 | iteration number: 225
best profit: 0.2999999999999998 | iteration number: 226
best profit: 0.2999999999999998 | iteration number: 227
best profit: 0.2999999999999998 | iteration number: 228
best profit: 0.2999999999999998 | iteration number: 229
best profit: 0.2999999999999998 | iteration number: 230
best profit: 0.2999999999999998 | iteration number: 231
best profit: 0.2999999999999998 | iteration number: 232
best profit: 0.2999999999999998 | iteration number: 233
best profit: 0.2999999999999998 | iteration number: 234
best profit: 0.2999999999999998 | iteration number: 235
best profit: 0.2999999999999998 | iteration numb

In [5]:
# Algorytm genetyczny, który w rozwiązaniach początkowych zawiera optymalną strategię zbiorów.
strategies = [
                [[1,[1,1]],[1,[1,1]]],
                [[1,[2,0]],[1,[0,2]]],
                [[1,[1,0]],[1,[0,1]]],
                [[1,[0,1]],[1,[1,0]]],
                [[1,[0,0]],[1,[2,0]]],
                [[1,[1,1]],[1,[2,0]]],
                [[1,[2,0]],[1,[1,1]]],
                [[1,[0,2]],[1,[1,1]]],
                [[1,[1,1]],[1,[0,2]]]
]
sol, profit, iterations = orchard.genetic_algorithm(max_iter_no_progress=300, max_iter=3000,
                                            replacement_rate=0.5,
                                            mutation_proba=0.4, random_demand_rate=False, verbose=True,
                                                   bruteforce_comapre=strategies)

print(f"Wynik algorytmu genetycznego\n{sol}Zysk: {profit}")

best profit: 0.7000000000000002 | iteration number: 1
best profit: 0.7000000000000002 | iteration number: 2
best profit: 0.7000000000000002 | iteration number: 3
best profit: 0.7000000000000002 | iteration number: 4
best profit: 0.7000000000000002 | iteration number: 5
best profit: 0.7000000000000002 | iteration number: 6
best profit: 0.7000000000000002 | iteration number: 7
best profit: 0.7000000000000002 | iteration number: 8
best profit: 0.7000000000000002 | iteration number: 9
best profit: 0.7000000000000002 | iteration number: 10
best profit: 0.7000000000000002 | iteration number: 11
best profit: 0.7000000000000002 | iteration number: 12
best profit: 0.7000000000000002 | iteration number: 13
best profit: 0.7000000000000002 | iteration number: 14
best profit: 0.7000000000000002 | iteration number: 15
best profit: 0.7000000000000002 | iteration number: 16
best profit: 0.7000000000000002 | iteration number: 17
best profit: 0.7000000000000002 | iteration number: 18
best profit: 0.7000

best profit: 0.7000000000000002 | iteration number: 204
best profit: 0.7000000000000002 | iteration number: 205
best profit: 0.7000000000000002 | iteration number: 206
best profit: 0.7000000000000002 | iteration number: 207
best profit: 0.7000000000000002 | iteration number: 208
best profit: 0.7000000000000002 | iteration number: 209
best profit: 0.7000000000000002 | iteration number: 210
best profit: 0.7000000000000002 | iteration number: 211
best profit: 0.7000000000000002 | iteration number: 212
best profit: 0.7000000000000002 | iteration number: 213
best profit: 0.7000000000000002 | iteration number: 214
best profit: 0.7000000000000002 | iteration number: 215
best profit: 0.7000000000000002 | iteration number: 216
best profit: 0.7000000000000002 | iteration number: 217
best profit: 0.7000000000000002 | iteration number: 218
best profit: 0.7000000000000002 | iteration number: 219
best profit: 0.7000000000000002 | iteration number: 220
best profit: 0.7000000000000002 | iteration numb

In [6]:
population = orchard.create_initial_population(bruteforce_comapre=strategies)
for sol in population:
    print(sol[0])
print(len(population))

Day 1
harvested: [1, 1]
sold_market: [0, 0]
sold_wholesale: [1, 1]
warehouse: [0, 0]

Day 2
harvested: [1, 1]
sold_market: [0, 0]
sold_wholesale: [1, 1]
warehouse: [0, 0]






Day 1
harvested: [1, 1]
sold_market: [1, 1]
sold_wholesale: [0, 0]
warehouse: [0, 0]

Day 2
harvested: [1, 1]
sold_market: [1, 1]
sold_wholesale: [0, 0]
warehouse: [0, 0]






Day 1
harvested: [1, 1]
sold_market: [0, 1]
sold_wholesale: [0, 0]
warehouse: [1, 0]

Day 2
harvested: [1, 1]
sold_market: [1, 1]
sold_wholesale: [0, 0]
warehouse: [1, 0]






Day 1
harvested: [2, 0]
sold_market: [0, 0]
sold_wholesale: [2, 0]
warehouse: [0, 0]

Day 2
harvested: [0, 2]
sold_market: [0, 0]
sold_wholesale: [0, 2]
warehouse: [0, 0]






Day 1
harvested: [2, 0]
sold_market: [1, 0]
sold_wholesale: [0, 0]
warehouse: [1, 0]

Day 2
harvested: [0, 2]
sold_market: [1, 2]
sold_wholesale: [0, 0]
warehouse: [0, 0]






Day 1
harvested: [2, 0]
sold_market: [0, 0]
sold_wholesale: [1, 0]
warehouse: [1, 0]

Day 2
harvested: [0, 2]
sold_

In [7]:
# Algorytm symulowanego wyżarzania, który startuje z rozwiązania innego niż optymalne.
sol, profit, _, _ = orchard.simulated_annealing(T_start=2000, T_stop=100, iterations_in_temp=100,
                                                        epsilon=2, iterations_epsilon=1000, alpha=0.99,
                                                        neighbour_type=1, initial_sol=3, verbose=True, bruteforce_comapre=strategies)
print(f"Wynik algorytmu wyrzarzania\n{sol}Zysk: {profit}")

best profit: -4.5 | temperature: 2000
best profit: -4.5 | temperature: 1980.0
best profit: -4.5 | temperature: 1960.2
best profit: -4.5 | temperature: 1940.598
best profit: -4.5 | temperature: 1921.19202
best profit: -4.5 | temperature: 1901.9800997999998
best profit: -4.5 | temperature: 1882.960298802
best profit: -4.5 | temperature: 1864.1306958139799
best profit: -4.5 | temperature: 1845.48938885584
best profit: -4.5 | temperature: 1827.0344949672815
best profit: -4.5 | temperature: 1808.7641500176087
best profit: -4.5 | temperature: 1790.6765085174327
best profit: -4.5 | temperature: 1772.7697434322583
best profit: -4.5 | temperature: 1755.0420459979357
best profit: -4.5 | temperature: 1737.4916255379562
best profit: -4.5 | temperature: 1720.1167092825767
best profit: -4.5 | temperature: 1702.915542189751
best profit: -4.5 | temperature: 1685.8863867678535
best profit: -4.5 | temperature: 1669.027522900175
best profit: -4.5 | temperature: 1652.3372476711731
best profit: -4.5 | temp

best profit: -4.5 | temperature: 388.65719776558973
best profit: -4.5 | temperature: 384.77062578793385
best profit: -4.5 | temperature: 380.9229195300545
best profit: -4.5 | temperature: 377.113690334754
best profit: -4.5 | temperature: 373.34255343140643
best profit: -4.5 | temperature: 369.6091278970924
best profit: -4.5 | temperature: 365.91303661812145
best profit: -4.5 | temperature: 362.25390625194024
best profit: -4.5 | temperature: 358.6313671894208
best profit: -4.5 | temperature: 355.0450535175266
best profit: -4.5 | temperature: 351.4946029823513
best profit: -4.5 | temperature: 347.97965695252776
best profit: -4.5 | temperature: 344.4998603830025
best profit: -4.5 | temperature: 341.0548617791725
best profit: -4.5 | temperature: 337.64431316138075
best profit: -4.5 | temperature: 334.2678700297669
best profit: -4.5 | temperature: 330.92519132946927
best profit: -4.5 | temperature: 327.61593941617457
best profit: -4.5 | temperature: 324.3397800220128
best profit: -4.5 | tem

In [8]:
# Algorytm symulowanego wyżarzania, który startuje z rozwiązania optymalnego.
sol, profit, _, _ = orchard.simulated_annealing(T_start=2000, T_stop=100, iterations_in_temp=100,
                                                        epsilon=2, iterations_epsilon=1000, alpha=0.99,
                                                        neighbour_type=1, initial_sol=25, verbose=True, bruteforce_comapre=strategies)
print(f"Wynik algorytmu wyrzarzania\n{sol}Zysk: {profit}")

best profit: 0.7000000000000002 | temperature: 2000
best profit: 0.7000000000000002 | temperature: 1980.0
best profit: 0.7000000000000002 | temperature: 1960.2
best profit: 0.7000000000000002 | temperature: 1940.598
best profit: 0.7000000000000002 | temperature: 1921.19202
best profit: 0.7000000000000002 | temperature: 1901.9800997999998
best profit: 0.7000000000000002 | temperature: 1882.960298802
best profit: 0.7000000000000002 | temperature: 1864.1306958139799
best profit: 0.7000000000000002 | temperature: 1845.48938885584
best profit: 0.7000000000000002 | temperature: 1827.0344949672815
best profit: 0.7000000000000002 | temperature: 1808.7641500176087
best profit: 0.7000000000000002 | temperature: 1790.6765085174327
best profit: 0.7000000000000002 | temperature: 1772.7697434322583
best profit: 0.7000000000000002 | temperature: 1755.0420459979357
best profit: 0.7000000000000002 | temperature: 1737.4916255379562
best profit: 0.7000000000000002 | temperature: 1720.1167092825767
best p

best profit: 0.7000000000000002 | temperature: 546.9783020444321
best profit: 0.7000000000000002 | temperature: 541.5085190239878
best profit: 0.7000000000000002 | temperature: 536.0934338337479
best profit: 0.7000000000000002 | temperature: 530.7324994954104
best profit: 0.7000000000000002 | temperature: 525.4251745004563
best profit: 0.7000000000000002 | temperature: 520.1709227554517
best profit: 0.7000000000000002 | temperature: 514.9692135278972
best profit: 0.7000000000000002 | temperature: 509.8195213926182
best profit: 0.7000000000000002 | temperature: 504.721326178692
best profit: 0.7000000000000002 | temperature: 499.6741129169051
best profit: 0.7000000000000002 | temperature: 494.67737178773604
best profit: 0.7000000000000002 | temperature: 489.7305980698587
best profit: 0.7000000000000002 | temperature: 484.83329208916007
best profit: 0.7000000000000002 | temperature: 479.98495916826846
best profit: 0.7000000000000002 | temperature: 475.1851095765858
best profit: 0.70000000

best profit: 0.7000000000000002 | temperature: 151.1036681350557
best profit: 0.7000000000000002 | temperature: 149.59263145370514
best profit: 0.7000000000000002 | temperature: 148.09670513916808
best profit: 0.7000000000000002 | temperature: 146.6157380877764
best profit: 0.7000000000000002 | temperature: 145.14958070689863
best profit: 0.7000000000000002 | temperature: 143.69808489982964
best profit: 0.7000000000000002 | temperature: 142.26110405083134
best profit: 0.7000000000000002 | temperature: 140.83849301032302
best profit: 0.7000000000000002 | temperature: 139.43010808021978
best profit: 0.7000000000000002 | temperature: 138.0358069994176
best profit: 0.7000000000000002 | temperature: 136.65544892942341
best profit: 0.7000000000000002 | temperature: 135.28889444012918
best profit: 0.7000000000000002 | temperature: 133.9360054957279
best profit: 0.7000000000000002 | temperature: 132.59664544077063
best profit: 0.7000000000000002 | temperature: 131.27067898636292
best profit: 0

Oba algorytmy nie były w stanie znaleźć optymalnego rozwiązania w momencie gdy wśród rozwiązań początkowych nie było rozwiązania optymalnego. Może to wynikać z faktu, że nawet dla tak prostego przypadku ilość dopuszczalnych rozwiązań jest bardzo duża.